components/function-controller/pkg/controller/function/function_controller.go

```diff
@@ -107,308 +100,360 @@ var (
 
 	// Get a new *RuntimeInfo
+	// TODO(antoineco): this validation should happen ahead of the reconciliation
 	rnInfo, err := runtimeUtil.New(fnConfig)
 	if err != nil {
-		log.Error(err, "Error while trying to get a new RuntimeInfo instance", "namespace", fnConfig.Namespace, "name", fnConfig.Name)
+		log.Error(err, "Error creating RuntimeInfo", "namespace", fnConfig.Namespace, "name", fnConfig.Name)
 		return reconcile.Result{}, err
 	}
 
-	// Create Function's ConfigMap
-	foundCm := &corev1.ConfigMap{}
-	deployCm := &corev1.ConfigMap{}
-	_, err = r.createFunctionConfigMap(foundCm, deployCm, fn)
+	// Synchronize Function ConfigMap
+	fnCm, err := r.syncFunctionConfigMap(fn)
 	if err != nil {
-		if errors.IsNotFound(err) {
-			return reconcile.Result{}, nil
+		if err := r.updateFunctionStatus(fn, serverlessv1alpha1.FunctionConditionError); err != nil {
+			log.Error(err, "Error setting Function status", "namespace", fn.Namespace, "name", fn.Name)
 		}
-		r.updateFunctionStatus(fn, serverlessv1alpha1.FunctionConditionError)
 
-		log.Error(err, "function configmap can't be created. The function could have been deleted.", "namespace", deployCm.Namespace, "name", deployCm.Name)
+		log.Error(err, "Error during sync of the Function ConfigMap", "namespace", fn.Namespace, "name", fn.Name)
 		return reconcile.Result{}, err
 	}
 
-	// Update Function's ConfigMap
-	if err := r.updateFunctionConfigMap(foundCm, deployCm); err != nil {
-		// status of the functon must change to error.
-		r.updateFunctionStatus(fn, serverlessv1alpha1.FunctionConditionError)
+	// Synchronize Function BuildTemplate
+	if err := r.syncFunctionBuildTemplate(fn, rnInfo); err != nil {
+		if err := r.updateFunctionStatus(fn, serverlessv1alpha1.FunctionConditionError); err != nil {
+			log.Error(err, "Error setting Function status", "namespace", fn.Namespace, "name", fn.Name)
+		}
 
-		log.Error(err, "Error while trying to update Function's ConfigMap:", "namespace", deployCm.Namespace, "name", deployCm.Name)
+		log.Error(err, "Error during sync of the Function BuildTemplate", "namespace", fn.Namespace, "name", buildTemplateName)
 		return reconcile.Result{}, err
 	}
 
-	// Create function's image name
-	hash := sha256.New()
-	hash.Write([]byte(foundCm.Data["handler.js"] + foundCm.Data["package.json"]))
-	functionSha := fmt.Sprintf("%x", hash.Sum(nil))
-	imageName := fmt.Sprintf("%s/%s-%s:%s", rnInfo.RegistryInfo, fn.Namespace, fn.Name, functionSha)
-	log.Info("function image", "namespace:", fn.Namespace, "name:", fn.Name, "imageName:", imageName)
-
-	if err := r.getFunctionBuildTemplate(fn, rnInfo); err != nil {
-		// status of the functon must change to error.
-		r.updateFunctionStatus(fn, serverlessv1alpha1.FunctionConditionError)
-
+	// Generate build and image names
+	fnSha, err := generateFunctionHash(fnCm)
+	if err != nil {
+		log.Error(err, "Error generating Function hash", "namespace", fn.Namespace, "name", fn.Name)
 		return reconcile.Result{}, err
 	}
+	shortSha := fnSha[:10]
```

Comment:
How about we replace hard coded value of 10 to some constant and with a note probably?

Reply:
done